gpt_index/vector_stores/redis.py

```diff
@@ -0,0 +1,313 @@
+            # as it will be created from the embedding result attributes.
+            self._vector_args["dims"] = len(embedding_results[0].embedding)
+            self._create_index()
+
+        ids = []
+        for result in embedding_results:
+            # Add extra info and node info to the index
+            node_info = result.node.node_info if result.node.node_info is not None else {}
+            extra_info = result.node.extra_info if result.node.extra_info is not None else {}
+
+            mapping = {
+                "id": result.id,
+                "doc_id": result.doc_id.replace("-", "_"),
+                "text": result.node.text,
+                "vector": np.array(result.embedding).astype(np.float32).tobytes(),
+                **node_info,
+                **extra_info
+            }
+            ids.append(result.id)
+            key = "_".join((self._prefix, str(result.id)))
+            self._redis_client.hset(key, mapping=mapping)
+
+        _logger.info(f"Added {len(ids)} documents to index {self._index_name}")
+        return ids
+
+
+    def delete(self, doc_id: str) -> None:
+        """Delete a specific document from the index by doc_id
+
+        Args:
+            doc_id (str): The doc_id of the document to delete.
+        """
+        # find all documents that match a doc_id
+        query_str = "@doc_id:{%s}" % doc_id.replace('-', '_')
+        results = self._redis_client.ft(self._index_name).search(query_str)
+        for doc in results.docs:
+            self._redis_client.delete(doc.id)
+        _logger.info(f"Deleted {len(results.docs)} documents from index {self._index_name}")
+
+
+    def delete_index(self):
+        """Delete the index and all documents."""
+        _logger.info(f"Deleting index {self._index_name}")
+        self._redis_client.ft(self._index_name).dropindex(delete_documents=True)
+
+    def query(self, query: VectorStoreQuery) -> VectorStoreQueryResult:
+        """Query the index.
+
+        Args:
+            query (VectorStoreQuery): query object
+
+        Returns:
+            VectorStoreQueryResult: query result
+        """
+        from redis.exceptions import ResponseError as RedisResponseError
+
+        redis_query = self._prep_query(query.similarity_top_k)
+        query_params = {"vector": np.array(query.query_embedding).astype(np.float32).tobytes()}
+        _logger.info(f"Querying index {self._index_name}")
+
+        try:
+            results = self._redis_client.ft(self._index_name).search(redis_query,
+                                                                    query_params=query_params)
+        except RedisResponseError as e:
+            _logger.error(f"Error querying index {self._index_name}: {e}")
+            raise e
+
+        ids = []
+        nodes = []
+        scores = []
+        for doc in results.docs:
+            node = Node(
+                text=doc.text,
+                doc_id=doc.doc_id.replace("_", "-"),
```

Comment:
this is super confusing and our fault, but `Node.doc_id` is the ID of the node, `Node.ref_doc_id` is the source document ID (set via `{DocumentRelationship.SOURCE: doc.doc_id}`).

So here we should restore with `doc_id=doc.id` I believe.